src/irmin-pack/unix/dispatcher.ml

```diff
@@ -85,11 +86,32 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
     let open Int63.Syntax in
     if off >= suffix_start_offset t then Some (soff_of_offset t off) else None
 
-  let read_exn t ~off ~len buf =
+  let read_exn t ~off ~len ?volume_identifier buf =
     [%log.debug "read_exn ~off:%a ~len:%i" Int63.pp off len];
+    let read_lower_exn ?volume () =
+      let lower = Fm.lower t.fm in
+      match lower with
+      | None -> None
+      | Some l -> Some (Lower.read_exn l ~off ~len ?volume buf)
+    in
+    let read_sparse_then_lower_exn () =
+      match Sparse.read_exn (get_prefix t) ~off ~len buf with
+      | exception (Errors.Pack_error (`Invalid_sparse_read (`Before, _)) as exn)
```

Comment:
I think the case `Invalid_sparse_read Hole` must also be handled here (it corresponds to reading an object that was live in the past and got garbaged collected out of the prefix, even though some even older objects survived longer than it... so we must go to the lower to read now)

Reply:
Ooh yes - good catch!